commit cd83cb953606b94966981056e79dbb6c48751055
Author: Andres Freund <andres@anarazel.de>
Date:   Fri Feb 25 10:30:05 2022 -0800

    pg_waldump: Fix error message for WAL files smaller than XLOG_BLCKSZ.
    
    When opening a WAL file smaller than XLOG_BLCKSZ (e.g. 0 bytes long) while
    determining the wal_segment_size, pg_waldump checked errno, despite errno not
    being set by the short read. Resulting in a bogus error message.
    
    Author: Kyotaro Horiguchi <horikyota.ntt@gmail.com>
    Discussion: https://postgr.es/m/20220214.181847.775024684568733277.horikyota.ntt@gmail.com
    Backpatch: 11-, the bug was introducedin fc49e24fa

diff --git a/src/bin/pg_waldump/pg_waldump.c b/src/bin/pg_waldump/pg_waldump.c
index a6251e1a961..2340dc247b0 100644
--- a/src/bin/pg_waldump/pg_waldump.c
+++ b/src/bin/pg_waldump/pg_waldump.c
@@ -220,19 +220,16 @@ search_directory(const char *directory, const char *fname)
 				fatal_error(ngettext("WAL segment size must be a power of two between 1 MB and 1 GB, but the WAL file \"%s\" header specifies %d byte",
 									 "WAL segment size must be a power of two between 1 MB and 1 GB, but the WAL file \"%s\" header specifies %d bytes",
 									 WalSegSz),
 							fname, WalSegSz);
 		}
+		else if (r < 0)
+			fatal_error("could not read file \"%s\": %m",
+						fname);
 		else
-		{
-			if (errno != 0)
-				fatal_error("could not read file \"%s\": %m",
-							fname);
-			else
-				fatal_error("could not read file \"%s\": read %d of %d",
-							fname, r, XLOG_BLCKSZ);
-		}
+			fatal_error("could not read file \"%s\": read %d of %d",
+						fname, r, XLOG_BLCKSZ);
 		close(fd);
 		return true;
 	}
 
 	return false;
